pkg/apply/filter/dependency-filter.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2020 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package filter
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/cli-utils/pkg/apply/taskrunner"
+	"sigs.k8s.io/cli-utils/pkg/object"
+)
+
+//go:generate stringer -type=Action -linecomment
+type Action int
+
+const (
+	ApplyAction  Action = iota // Apply
+	DeleteAction               // Delete
+)
+
+// DependencyFilter implements ValidationFilter interface to determine if an
+// object can be applied or deleted based on the status of it's dependencies.
+type DependencyFilter struct {
+	TaskContext *taskrunner.TaskContext
+	Action      Action
+}
+
+const DependencyFilterName = "DependencyFilter"
+
+// Name returns the name of the filter for logs and events.
+func (dnrf DependencyFilter) Name() string {
+	return DependencyFilterName
+}
+
+// Filter returns true if the specified object should be skipped because at
+// least one of its dependencies is Not Found or Not Reconciled.
+func (dnrf DependencyFilter) Filter(obj *unstructured.Unstructured) (bool, string, error) {
+	id := object.UnstructuredToObjMetadata(obj)
+
+	switch dnrf.Action {
+	case ApplyAction:
+		edgesFrom := dnrf.TaskContext.Graph.EdgesFrom(id)
+		for _, depID := range edgesFrom {
+			if dnrf.TaskContext.IsInvalidObject(depID) {
+				return true, fmt.Sprintf("dependency invalid: %q", depID), nil
+			}
+			if dnrf.TaskContext.IsSkippedApply(depID) {
+				return true, fmt.Sprintf("dependency apply skipped: %q", depID), nil
+			}
+			if dnrf.TaskContext.IsFailedApply(depID) || !dnrf.TaskContext.IsSuccessfulApply(depID) {
```

Comment:
It feels redundant to have both `IsFailedApply` and `IsSuccessfulApply` but they're not always one or the other... There's some weird code in the applier that might cause neither to be true.

Reply:
Same is true for delete and reconcile status caches in TaskContext...
Maybe we just need to simplify into one status for each: apply, delete, reconcile?
Or maybe just apply & reconciles, with both meanings including apply and delete?